File: lectures/lecture-01-why-capable-agents-still-fail/exercises/exercise-02-verification-gap/solution/typescript/main.ts
// verification-gap exercise, TypeScript solution.
//
// Classifies every run by whether its first claim was backed by an earlier
// passing verification, then computes the verification gap (unverified
// claims over all claims). Contract: ../../SPEC.md.

import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

interface RunEvent {
  readonly run: string;
  readonly type: string;
  readonly detail: string;
  readonly result?: string;
}

interface RunClassification {
  readonly claimed: boolean;
  readonly verified_before_claim: boolean;
  readonly classification: "verified-done" | "unverified-done" | "no-claim";
}

export function classify(events: readonly RunEvent[]): RunClassification {
  const firstClaimIndex = events.findIndex((e) => e.type === "claim");
  if (firstClaimIndex === -1) {
    return { claimed: false, verified_before_claim: false, classification: "no-claim" };
  }
  const verified = events
    .slice(0, firstClaimIndex)
    .some((e) => e.type === "verification" && e.result === "pass");
  return {
    claimed: true,
    verified_before_claim: verified,
    classification: verified ? "verified-done" : "unverified-done",
  };
}

export function gapReport(events: readonly RunEvent[]): Record<string, unknown> {
  const order: string[] = [];
  const runs = new Map<string, RunEvent[]>();
  for (const event of events) {
    let entry = runs.get(event.run);
    if (!entry) {
      entry = [];
      runs.set(event.run, entry);
      order.push(event.run);
    }
    entry.push(event);
  }

  const reportRuns = [];
  let claims = 0;
  let verifiedClaims = 0;
  for (const runId of order) {
    const result = classify(runs.get(runId) ?? []);
    reportRuns.push({ id: runId, ...result });
    if (result.claimed) {
      claims += 1;
      if (result.verified_before_claim) verifiedClaims += 1;
    }
  }

  const unverified = claims - verifiedClaims;
  return {
    runs: reportRuns,
    claims,
    verified_claims: verifiedClaims,
    unverified_claims: unverified,
    verification_gap: claims > 0 ? unverified / claims : 0,
  };
}

export function parseTranscript(text: string): RunEvent[] {
  const events: RunEvent[] = [];
  const lines = text.split(/\r?\n/);
  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index];
    if (!line || !line.trim()) continue;
    let event: RunEvent;
    try {
      event = JSON.parse(line) as RunEvent;
    } catch (error) {
      throw new Error(`malformed transcript at line ${index + 1}: ${String(error)}`, {
        cause: error,
      });
    }
    for (const field of ["run", "type", "detail"] as const) {
      if (!(field in event)) {
        throw new Error(`malformed transcript at line ${index + 1}: missing '${field}'`);
      }
    }
    events.push(event);
  }
  return events;
}

function main(argv: readonly string[]): number {
  const path = argv[2];
  if (!path || argv.length !== 3) {
    console.error("usage: main.ts <transcript.jsonl>");
    return 2;
  }
  let text: string;
  try {
    text = readFileSync(path, "utf8");
  } catch (error) {
    console.error(`error: cannot read transcript: ${String(error)}`);
    return 2;
  }
  let events: RunEvent[];
  try {
    events = parseTranscript(text);
  } catch (error) {
    console.error(`error: ${error instanceof Error ? error.message : String(error)}`);
    return 1;
  }
  process.stdout.write(JSON.stringify(gapReport(events), null, 2) + "\n");
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}

File: lectures/lecture-01-why-capable-agents-still-fail/exercises/exercise-02-verification-gap/solution/python/main.py
"""verification-gap exercise, Python solution.

Classifies every run by whether its first claim was backed by an earlier
passing verification, then computes the verification gap (unverified claims
over all claims). Contract: ../../SPEC.md.
"""

from __future__ import annotations

import json
import sys


def classify(events: list[dict]) -> dict:
    """Classify one run's events (in execution order)."""
    first_claim_index = next(
        (i for i, e in enumerate(events) if e["type"] == "claim"), None
    )
    if first_claim_index is None:
        return {"claimed": False, "verified_before_claim": False, "classification": "no-claim"}
    verified = any(
        e["type"] == "verification" and e.get("result") == "pass"
        for e in events[:first_claim_index]
    )
    return {
        "claimed": True,
        "verified_before_claim": verified,
        "classification": "verified-done" if verified else "unverified-done",
    }


def gap_report(events: list[dict]) -> dict:
    order: list[str] = []
    runs: dict[str, list[dict]] = {}
    for event in events:
        run_id = event["run"]
        if run_id not in runs:
            order.append(run_id)
            runs[run_id] = []
        runs[run_id].append(event)

    report_runs = []
    claims = verified_claims = 0
    for run_id in order:
        entry = {"id": run_id, **classify(runs[run_id])}
        report_runs.append(entry)
        if entry["claimed"]:
            claims += 1
            if entry["verified_before_claim"]:
                verified_claims += 1

    unverified = claims - verified_claims
    return {
        "runs": report_runs,
        "claims": claims,
        "verified_claims": verified_claims,
        "unverified_claims": unverified,
        "verification_gap": unverified / claims if claims else 0.0,
    }


def parse_transcript(text: str) -> list[dict]:
    events = []
    for number, line in enumerate(text.split("\n"), 1):
        if not line.strip():
            continue
        try:
            event = json.loads(line)
        except json.JSONDecodeError as error:
            raise ValueError(f"malformed transcript at line {number}: {error}") from error
        for field in ("run", "type", "detail"):
            if field not in event:
                raise ValueError(f"malformed transcript at line {number}: missing {field!r}")
        events.append(event)
    return events


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <transcript.jsonl>", file=sys.stderr)
        return 2
    try:
        with open(argv[1], encoding="utf-8") as handle:
            text = handle.read()
    except OSError as error:
        print(f"error: cannot read transcript: {error}", file=sys.stderr)
        return 2
    try:
        events = parse_transcript(text)
    except ValueError as error:
        print(f"error: {error}", file=sys.stderr)
        return 1
    print(json.dumps(gap_report(events), indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
